pkg/querier/queryrange/limits_test.go

```diff
@@ -0,0 +1,218 @@
+			expectedSkipped:  true,
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			req := &PrometheusRequest{
+				Start: util.TimeToMillis(testData.reqStartTime),
+				End:   util.TimeToMillis(testData.reqEndTime),
+			}
+
+			limits := mockLimits{maxQueryLookback: testData.maxQueryLookback}
+			middleware := NewLimitsMiddleware(limits)
+
+			innerRes := NewEmptyPrometheusResponse()
+			inner := &mockHandler{}
+			inner.On("Do", mock.Anything, mock.Anything).Return(innerRes, nil)
+
+			ctx := user.InjectOrgID(context.Background(), "test")
+			outer := middleware.Wrap(inner)
+			res, err := outer.Do(ctx, req)
+			require.NoError(t, err)
+
+			if testData.expectedSkipped {
+				// We expect an empty response, but not the one returned by the inner handler
+				// which we expect has been skipped.
+				assert.NotSame(t, innerRes, res)
+				assert.Len(t, inner.Calls, 0)
+			} else {
+				// We expect the response returned by the inner handler.
+				assert.Same(t, innerRes, res)
+
+				// Assert on the time range of the request passed to the inner handler (5s delta).
+				delta := float64(5000)
+				require.Len(t, inner.Calls, 1)
+				assert.InDelta(t, util.TimeToMillis(testData.expectedStartTime), inner.Calls[0].Arguments.Get(1).(Request).GetStart(), delta)
+				assert.InDelta(t, util.TimeToMillis(testData.expectedEndTime), inner.Calls[0].Arguments.Get(1).(Request).GetEnd(), delta)
+			}
+		})
+	}
+}
+
+func TestLimitsMiddleware_MaxQueryLength(t *testing.T) {
+	const (
+		thirtyDays = 30 * 24 * time.Hour
+	)
+
+	now := time.Now()
+
+	tests := map[string]struct {
+		maxQueryLength time.Duration
+		reqStartTime   time.Time
+		reqEndTime     time.Time
+		expectedErr    string
+	}{
+		"should skip validation if max length is disabled": {
+			maxQueryLength: 0,
+			reqStartTime:   time.Unix(0, 0),
+			reqEndTime:     now,
+		},
+		"should succeed on a query on short time range, ending now": {
+			maxQueryLength: thirtyDays,
+			reqStartTime:   now.Add(-time.Hour),
+			reqEndTime:     now,
+		},
+		"should succeed on a query on short time range, ending in the past": {
+			maxQueryLength: thirtyDays,
+			reqStartTime:   now.Add(-2 * thirtyDays).Add(-time.Hour),
+			reqEndTime:     now.Add(-2 * thirtyDays),
+		},
+		"should succeed on a query on large time range close to the limit, ending now": {
+			maxQueryLength: thirtyDays,
+			reqStartTime:   now.Add(-thirtyDays).Add(time.Hour),
+			reqEndTime:     now,
+		},
+		"should fail on a query on large time range over the limit, ending now": {
+			maxQueryLength: thirtyDays,
+			reqStartTime:   now.Add(-thirtyDays).Add(-100 * time.Hour),
+			reqEndTime:     now,
+			expectedErr:    "the query time range exceeds the limit",
+		},
+		"should fail on a query on large time range over the limit, ending in the past": {
+			maxQueryLength: thirtyDays,
+			reqStartTime:   now.Add(-2 * thirtyDays).Add(-thirtyDays).Add(-100 * time.Hour),
```

Comment:
This is getting hard to read. What about this?

```suggestion
			reqStartTime:   now.Add(-3*thirtyDays - 100*time.Hour),
```